[Feature] Command Argument STDOUT/capistrano.log Hiding

NorseGaud opened the pull request:
Take the following task in capistrano-postgresql for example:
```
desc 'Create pg_username in database'
  task :create_database_user do
    on roles :db do
      unless database_user_exists?
        # If you use CREATE USER instead of CREATE ROLE the LOGIN right is granted automatically; otherwise you must specify it in the WITH clause of the CREATE statement.
        execute [:psql, "-d #{fetch(:pg_system_db)}", '-c', %Q{"CREATE USER \\"#{fetch(:pg_username)}\\" PASSWORD '}, ["#{fetch(:pg_password)}"], %Q{';"}]
      end
    end
  end
```

The wrapping of "#{fetch(:pg_password)}" with [ ] will allow us to hide the password the :pretty log STDOUT and capistrano.log entry:

```00:00 postgresql:create_database_user
      01 sudo -i -u postgres psql -d postgres -c "CREATE USER \"growtrader\" PASSWORD ' *HIDDEN* ';"
      01 CREATE ROLE
    ✔ 01 growtrader@growtrader.dev 0.096s
```
``` INFO [600d7e15] Running /usr/bin/env sudo -i -u postgres psql -d postgres -c "CREATE USER \"growtrader\" PASSWORD ' *HIDDEN* ';" as growtrader@growtrader.dev
 DEBUG [600d7e15] Command: ( export PATH="$HOME/.gem/ruby/2.5.0/bin:$PATH" ; /usr/bin/env sudo -i -u postgres psql -d postgres -c "CREATE USER \"growtrader\" PASSWORD ' *HIDDEN* ';" )
 DEBUG [c4a860dc] 	CREATE ROLE
  INFO [c4a860dc] Finished in 0.096 seconds with exit status 0 (successful).
```

Yet, server side, I see 

```[growtrader@growtrader ~]$ cat db/database.yml
development:
  adapter: postgresql
  encoding: UTF-8
  database: growtrader
  pool: 100
  username: growtrader
  password: eaf38e7bec5d860c44d5
  host: growtrader.dev
  socket:
  port: 5432
  timeout: 5000
```

1. It supports multiple strings in the args array you pass in.
2. It removed the args array (using join, just in case the user has multiple things in the hidden array) before executing it (current tests, except for the vagrant stuff, pass)
3. I will create tests and also document it once I get sign off on the current code. It's 12:11AM right now, so I need to sleep. :)
4. RuboCop only has one offense, and it's minor, and I can include that with the readme and tests I add if this is approved.
5. I'll add the Changelog entry once approved.

Look forward to your feedback!

NorseGaud commented:
Travis failures look to be from RuboCop. I'll fix those after the code is approved.

leehambley commented:
Really nice approach, of course as you alluded to it needs tests/etc but it's workable.

I wonder if another approach might be cleaner and/or give a better developer experience. Namely that we don't abuse `[]` arrays but that we do something like using Ruby's [Refinements](https://docs.ruby-lang.org/en/2.4.0/syntax/refinements_rdoc.html) feature to mark something as `.redactable` or similar.

I don't want to bog you down with more research, and I really really appreciate your contribution, the array idea is really elegant to single out elements for special treatment, I just don't know how well it works in the API as a whole.

leehambley commented:
I just tried (and failed) to make something with refinements work, but I think it's because I'm shooting from the hip instead of doing it properly.

It just came to my mind though that Ruby has `Object.trust`, `Object.taint` and `Object.trusted?`, `Object.tainted?` etc - these depend a bit on env vars, but we might have success with something like:

```
unless database_user_exists?
        # If you use CREATE USER instead of CREATE ROLE the LOGIN right is granted automatically; otherwise you must specify it in the WITH clause of the CREATE statement.
        execute [:psql, "-d #{fetch(:pg_system_db)}", '-c', %Q{"CREATE USER \\"#{fetch(:pg_username)}\\" PASSWORD '}, "#{fetch(:pg_password)}".taint, %Q{';"}]
      end
```

and then simply use `.tainted?` in the logger/formatter.

NorseGaud commented:
Oooo, yeah! I like that direction a lot. I'll play around with it in my developer environment and see if I can get it to work. Thanks for the suggestion!

leehambley commented:
Cool! The idea with refinements would be similar, to add a ".redacted"
trait/refinement, but I don't know if they affects all strings etc

Sent from my phone from underway.

On Fri, 8 Jun 2018, 14:46 Nathan Pierce, <notifications@github.com> wrote:

> Oooo, yeah! I like that direction a lot. I'll play around with it in my
> developer environment and see if I can get it to work. Thanks for the
> suggestion!
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/capistrano/sshkit/pull/430#issuecomment-395750035>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AABKCGn4REsDGqkpHWypeyzDtQGRRDJSks5t6nIXgaJpZM4Ufd6R>
> .
>

mattbrictson commented:
Capistrano currently supports Ruby 2.0 and higher, and I think refinements didn't fully settle until 2.1. I'm not sure I'd want to bump the Ruby requirement just for this feature, so it might be best to avoid using refinements in the implementation.

NorseGaud commented:
Yep, experimental in 2.0.0. Would the code I’ve written be ok for now, until we eventually bump the ruby version requirement? I can put a note in the code about it.

leehambley commented:
Good catch @mattbrictson I always forget how much work you put into the backwards compatibility.

Can you live with the "misuse" of trusted/tainted or similar to avoid the version bump, or do you have a better idea?

will-in-wi commented:
Do we have a more or less official backwards compatibility policy? As of April of last year, 2.1 is fully unsupported. We may want to take this opportunity to drop support for 2.0 if it means cleaner code.

leehambley commented:
> Do we have a more or less official backwards compatibility policy? As of April of last year, 2.1 is fully unsupported. We may want to take this opportunity to drop support for 2.0 if it means cleaner code.

Either time is accelerating, or I'm getting old. I remember *looking forward* to Ruby 2.x like it was the bright new future.

I'd throw a vote for our policy being "anything that is not supported/EOL'ed upstream" being our target.

NorseGaud commented:
I'll go ahead and make the necessary remaining changes and also use tainted/etc like you suggested Lee.

NorseGaud commented:
Ok, I think the new changes are user-friendly enough and I've also confirmed the tests in vagrant are passing. I tested it in my own project as well and it's working well.

What do you think @leehambley?

leehambley commented:
The user-facing API is super elegant, and a wrapper method avoids all the issues with Ruby versions, and how Refinements, etc all work, so I applaud you for that. 

The slight abuse or arrays still makes me a little bit uncomfortable, but now that it's hidden behind a method call I'm ok with it. I'd like to see some const like `REDACTION_STRING = "*REDACTED*"` and a check in the logger around `arg.is_a?(Array) ? arg[0] : arg` also to check that `arg[0]` actually is the redaction string, just to avoid that someone accidentally passes an array and gets really hard to debug behaviour.

With all that said, from my side, this is good to go if we get that extra level of safety in. I hope the others would agree.

I want to thank you for the best FOSS contributorship experience I've had in a long time, this has been fun.

NorseGaud commented:
Glad to hear it Lee! My pleasure man. I really love capistrano and appreciate the warm welcome by the community. 

I totally agree with the elimination of the [redacted,arg] array in the redact method. I'm sort of out of ideas since we can't use .taint. If you have any suggestions on how to accomplish it, let me know!

In the meantime, I've added a check to raise if the user tries to pass in redact([]) or redact({}). I've added to the test_redaction a simple check for the exception as well.

mattbrictson commented on CHANGELOG.md:
```diff
@@ -7,6 +7,10 @@ appear at the top.
 
+## [1.17.0][] (2018-06-08)
+
+  * [#430](https://github.com/capistrano/sshkit/pull/430): [Feature] Command Argument STDOUT/capistrano.log Hiding - [@NorseGaud](https://github.com/NorseGaud)
```
Please move this to the Unreleased section (next to the "Your contribution here!" bullet).

mattbrictson commented on lib/sshkit/version.rb:
```diff
@@ -1,3 +1,3 @@
 module SSHKit
-  VERSION = "1.16.1".freeze
+  VERSION = "1.17.0".freeze
```
Sorry, only maintainers are allowed to change the version 😅

NorseGaud replied:
No problemo!

mattbrictson commented on test/functional/backends/test_netssh.rb:
```diff
@@ -42,13 +42,31 @@ def test_simple_netssh
         ], command_lines
       end
 
+      def test_redaction
```
Nice test! 💯

mattbrictson commented on lib/sshkit/backends/netssh.rb:
```diff
@@ -123,7 +129,12 @@ def transfer_summarizer(action, options = {})
+
+        # Redaction
+        cmd_w_redaction = cmd.dup
+        cmd.dup.args.map!{|arg| arg.is_a?(Array) ? arg[0] : arg }
```
This does redact the arguments from the logs (as verified by the unit test), but I suspect it is also redacting (and thus breaking) the actual command that is being executed.

The issue is that `dup` only makes a shallow copy. Therefore `cmd.dup.args` is the same reference as `cmd.args`, and so `args.map!` is mutating both copies of the command. Furthermore, don't we want to log `arg[0]` (i.e. the `*REDACTED*` string), but execute `arg[1]`? I don't see where the latter is taking place in the code.

I tested this branch using a Capistrano task:

```ruby
task :test_redact do
  on roles(:all) do
    execute :echo, redact("secret!"), "> test_redact"
    puts capture(:cat, "test_redact") #  this should print `secret!`
  end
end
```

And confirmed that `secret!` is being redacted from the log, but the original `secret!` is never actually passed to the command being executed:

```
00:00 test_redact
      01 echo *REDACTED* > test_redact
    ✔ 01 deployer@xx.xx.xx.xx 2.344s
*REDACTED*
```

NorseGaud replied:
Hmm, ok, I’ll run more tests. I had manually confirmed it was working but could have just confused myself into thinking it was fine

mattbrictson commented on .gitignore:
```diff
@@ -6,3 +6,4 @@ bin/rake
 .vagrant*
 test/tmp
 Gemfile.lock
+.idea
```
I'd rather not put IDE-specific patterns in our gitignore. Those are more appropriate for a global gitignore (see <https://thomashunter.name/blog/global-gitignore-vs-repository-gitignore/>)

mattbrictson commented on README.md:
````diff
@@ -443,6 +443,33 @@ SSHKit.config.output = SSHKit::Formatter::Pretty.new(output)
 SSHKit.config.output = SSHKit::Formatter::SimpleText.new(File.open('log/deploy.log', 'wb'))
 ```
 
+#### Output & Log Redaction
````
Yes, documentation!! ❤️

leehambley commented:
Anything standing in the way of merging this @mattbrictson , @will-in-wi

NorseGaud commented:
Yep, Matt is right about it not working fully. I’m testing a new way of doing it. Should be done by next week.

NorseGaud commented:
@leehambley you or @mattbrictson getting these weird vagrant/test errors?

`test_upload_large_file                                         ERROR (0.04s)
Net::SSH::HostKeyUnknown:         Net::SSH::HostKeyUnknown: fingerprint SHA256:I2efvQ+daDNq+OLVw9uThbTBKiu63nB0Qf+geS7cggo is unknown for "[localhost]:3001,[127.0.0.1]:3001"
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/verifiers/always.rb:46:in `process_cache_miss'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/verifiers/always.rb:21:in `verify'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/kex/diffie_hellman_group1_sha1.rb:188:in `verify_server_key'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/kex/diffie_hellman_group1_sha1.rb:72:in `exchange_keys'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/algorithms.rb:389:in `exchange_keys'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/algorithms.rb:211:in `proceed!'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/algorithms.rb:150:in `accept_kexinit'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:210:in `block in poll_message'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:190:in `loop'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:190:in `poll_message'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:225:in `block in wait'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:223:in `loop'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:223:in `wait'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh/transport/session.rb:90:in `initialize'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh.rb:242:in `new'
            /usr/local/lib/ruby/gems/2.5.0/gems/net-ssh-5.0.1/lib/net/ssh.rb:242:in `start'
            /Users/norsegaud/RubymineProjects/sshkit/lib/sshkit/backends/connection_pool.rb:59:in `call'
            /Users/norsegaud/RubymineProjects/sshkit/lib/sshkit/backends/connection_pool.rb:59:in `with'
            /Users/norsegaud/RubymineProjects/sshkit/lib/sshkit/backends/netssh.rb:184:in `with_ssh'
            /Users/norsegaud/RubymineProjects/sshkit/lib/sshkit/backends/netssh.rb:67:in `upload!'
            /Users/norsegaud/RubymineProjects/sshkit/test/functional/backends/test_netssh.rb:193:in `block in test_upload_large_file'
            /Users/norsegaud/RubymineProjects/sshkit/lib/sshkit/backends/abstract.rb:29:in `instance_exec'
            /Users/norsegaud/RubymineProjects/sshkit/lib/sshkit/backends/abstract.rb:29:in `run'
            /Users/norsegaud/RubymineProjects/sshkit/test/functional/backends/test_netssh.rb:195:in `test_upload_large_file'
`

NorseGaud commented:
Ok, passing all test in Travis and confirmed it works (after fixing my postgresql issue locally) in capistrano-postgresql.

Let me know if that covers every request.

mattbrictson commented:
Thanks! I'll look at it soon, probably over the weekend.

NorseGaud commented:
No rush :)

mattbrictson commented on lib/sshkit/backends/netssh.rb:
```diff
@@ -96,6 +96,12 @@ def config
+
+      def redact(arg) # Used in execute_command to hide redact() values user passes in
+        raise ArgumentError, 'redact() does not support Array or Hash' if arg.is_a?(Array) || arg.is_a?(Hash)
+        ['*REDACTED*',arg]
```
This works and I am happy to accept this PR as-is, but there might be a more elegant solution if you want to explore it. I am thinking of using a module as a way to "mark" the arg being redacted. In other words:

```ruby
module Redacted
  # intentionally empty
end
```

Then the `redact` method becomes:

```ruby
def redact(arg)
  arg.extend(Redacted)
end
```

Finally, testing whether something is redacted is:

```ruby
arg.is_a?(Redacted)
```

The advantage to this approach is that the arguments are modified in a way that should have no side-effects. Whereas replacing a string with a specially-formed array inherently modifies the arguments in a way that has to explicitly reversed later.

mattbrictson replied:
I anticipate that some SSHKit users will complain that `redact` works for SSH backends but not for local backends. Generally speaking we try to keep the two in parity. Is there an easy way to add the `redact` functionality to the local backend as well?

mattbrictson commented on lib/sshkit/backends/netssh.rb:
```diff
@@ -123,7 +129,9 @@ def transfer_summarizer(action, options = {})
       def execute_command(cmd)
+        cmd.options[:redacted] = true
         output.log_command_start(cmd)
+        cmd.options[:redacted] = false
```
This code strikes me as being a bit brittle. We are setting an option purely to get a desirable side-effect, when a more direct approach would be more self-explanatory.

What if the Command object had a method to get a redacted version of itself, and that is what gets logged?

I am open to suggestions on the name, but one possibility is:

```ruby
output.log_command_start(cmd.with_sensitive_args_redacted)
```

The `with_sensitive_args_redacted` method would be responsible for returning a copy of the command with a different set of args.

```ruby
def with_sensitive_args_redacted
  new_args = args.map( ... ) # logic to replace redacted args goes here
  redacted_cmd = dup
  redacted_cmd.instance_variable_set (:@args, new_args)
  redacted_cmd
end
```

mattbrictson commented on lib/sshkit/command.rb:
```diff
@@ -206,7 +206,14 @@ def to_command
       if should_map?
-        [SSHKit.config.command_map[command.to_sym], *Array(args)].join(' ')
+        if options[:redacted]
+          new_args = args.map{|arg| arg.is_a?(Array) && arg[0] == '*REDACTED*' ? arg[0] : arg }.join(' ')
```
I don't think the `join(' ')` here is necessary

NorseGaud commented:
Total rewrite using suggestions from @mattbrictson. I LOVE it. It's so simple now. Thank you sir!

- Module Redaction under SSHKit, so it's available everywhere we need it to be
- Simplified redact and moved it to abstact.rb so we can use it where we need
- with_redaction method available for SSHKit::Command
- Added .with_redaction to netssh backends, log_command_start
- Fixed tests, added Array, Hash, etc
- Rewrote README

mattbrictson commented:
@leehambley @will-in-wi I'll merge this if there are no further suggestions?

leehambley commented:
I'm very happy with this, the empty `is_a` is elegant in ways that give me the heebie jeebies, but I'm good with it, good shout with that approach.

mattbrictson commented:
Congrats on your first contribution to the project, @NorseGaud 🎉

NorseGaud commented:
Thanks again guys! What an enjoyable learning experience for me!

NorseGaud commented:
Quick question: what’s the next steps for getting the new gem version released? Anything I have to do or request?

mattbrictson commented:
We're on roughly a six-week release cycle. The next version is due out in the upcoming week and will include this PR. Nothing you need to do 🙂
